### DISCUSSION TITLE
Is dead-letter supported in in-memory transport?

### DISCUSSION THREAD

**Discussion by kalluu-ck** [celery/kombu#1508] [Q&A] [answered]
Currently, I want to implement unit tests for my error handler in dead-letter queue in Kombu.

With in-memory transport in Kombu, is the message routed to dead-letter queue when it is rejected?

**Comment by TayfurYldz** [comment, marked as answer]
Not in the RabbitMQ sense of "reject this consumed message and route it through a DLX".

The in-memory transport uses Kombu's virtual transport. Its QoS reject path is essentially:

```python
def reject(self, delivery_tag, requeue=False):
    if requeue:
        self.channel._restore_at_beginning(self._delivered[delivery_tag])
    self._quick_ack(delivery_tag)
```

So `reject(..., requeue=False)` removes the delivered message; it does not republish that rejected message to a dead-letter exchange/queue.

There *is* a `deadletter_queue` option in the virtual transport, but that's for otherwise **undeliverable routing**: `Channel._lookup()` falls back to that queue when no binding matches. It is not equivalent to RabbitMQ's dead-letter-exchange behavior for rejected/expired messages.

I checked the current Kombu source before answering: `kombu.transport.memory.Transport` inherits the virtual transport, and the virtual QoS reject implementation only restores on `requeue=True` or acknowledges/removes otherwise.

For a unit test whose purpose is specifically to exercise broker DLX semantics after `reject(requeue=False)`, use a broker/transport that actually implements those semantics (for example a disposable RabbitMQ instance). If you only want to unit-test your error-handler function, keep that as a separate test and feed it the dead-letter-shaped message directly; that avoids making the test depend on behavior the memory transport doesn't emulate.
